fix: HTTP response with invalid headers doesn't throw error #28865

BernardoSousa03 opened the pull request:
When receiving the described HTTP response Cypress resets the headers. This would cause the validateHeaderName method from node to be called which would cause an error, since the headers where invalid. Now Crypress verifies all the headers before reseting them, discards invalid ones and sends a warning in the console when debug module is on.

<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes https://github.com/cypress-io/cypress/issues/28865

### Additional details
The error lead to Cypress throwing errors when it wasn't supposed to

### Steps to test
Create a server that responds with invalid headers to a HTTP request

### How has the user experience changed?
The tests don't fail anymore and with the debug module on there is a warning that shows the invalid headers

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [x] Have tests been added/updated?
- [N/A] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [N/A] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

BernardoSousa03 committed:
fix: HTTP response with invalid headers doesn't throw error #28865

When receiving the described HTTP response Cypress resets the headers.
This would cause the validateHeaderName method from node to be called
which would cause an error, since the headers where invalid.
Now Crypress verifies all the headers before reseting them,
discards invalid ones and sends a warning in the console
when debug module is on.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cypress-io/cypress?pullRequest=29420) <br/>All committers have signed the CLA.

jennifer-shehane commented:
@BernardoSousa03 Thanks for the contribution and for adding a test case around the issue! I’ll check in the upcoming sprint to see who is available to review.

cacieprins commented:
When you consider the http proxy's role to sit seamlessly between the servers of the application under test, and the application under test, this makes sense. When you consider Cypress' role as an end-to-end testing platform, however, the failure makes sense: by not omitting malformed HTTP headers, Cypress helps ensure that the servers of the application under test are setting well-formed headers. Technically this is a breaking change, I think - it's fundamentally changing how Cypress responds to an error that originates with an external system.

Adding an experimental configuration option to warn when invalid headers are encountered might be the way to go here. Different teams may also have different opinions on whether they would want malformed headers to fail their tests, which also points to a configuration value. @jennifer-shehane , what do you think?

cacieprins reviewed (changes requested):
I think this approach is fine for now, especially considering the method of failure before. I'd like to see a warning logged to console rather than hiding it behind a debug statement, though - it's important that the modification of the headers be obvious, in case the client is expecting malformed headers.

BernardoSousa03 commented:
Thank you for your review @cacieprins. I'll get to it as soon as i can!

cacieprins commented:
Great, thank you @BernardoSousa03 !

Additionally, it may be useful to include validation of the header value, as well, with `validateHeaderValue` from the `http` package.

I recommend taking a look at `/packages/errors/`.

* a new entry can be added to `errors.ts` for templating how the error message is displayed in stdout
* including the error thrown, original http header (if not included in the error object's message), as well as the the URL and method being proxied for the warning display would be great
* a new entry in `packages/errors/test/unit/errors_spec.ts` for CI to accept the new error template - make sure to run `yarn test` in this package to generate a snapshot of the formatted error message. You can review the error message visually by opening the corresponding  HTML file in `./packages/errors/__snapshot-html__/`
* `errors.warning('PROXY_ENCOUNTERED_INVALID_HEADER_NAME', error, headerName)` and `errors.warning('PROXY_ENCOUNTERED_INVALID_HEADER_VALUE', error, headerValue)` would be the expected usage in the `proxy` package.

This message could read similar to:

```
Warning: While proxying a ${fmt.highlight(method)} request to ${fmt.url(urll)}, an HTTP header did not pass validation, and was removed. This header will not be present in the response received by the application under test.

Invalid header name: ${fmt.code(JSON.stringify(header, undefined, 2))}

${fmt.highlightSecondary(error)
```

BernardoSousa03 commented:
@cacieprins while trying out the `/packages/errors/` i found out that there is no visual representation for the warning while using `errors.warning` in the cypress GUI app. Is this what you had in mind?  I was thinking, since there is no plausible reason for having badly formed headers, it might be of interest to show the user it's mistake. I tried it with `errors.throwErr` but with no success. Should i just stick to using the first solution?

cacieprins commented:
The errors package is for displaying errors to the commandline - they won't show in the GUI, but they will be helpful for folks who may run into issues with this in CI. Please use errors.warning, instead of errors.throwErr, because we don't want to fully throw an error when this is encountered :)

BernardoSousa03 committed:
fix: improved warning display to the command line

When cutting off invalid headers from the response the user
is informed of such headers in the command line

BernardoSousa03 commented:
@cacieprins i have added the requested changes, anything else let me know!

cacieprins commented on packages/proxy/lib/http/response-middleware.ts:
```diff
@@ -306,7 +308,35 @@ const OmitProblematicHeaders: ResponseMiddleware = function () {
+        validateHeaderName(key)
+        if (Array.isArray(value)) {
+          value.forEach((v) => validateHeaderValue(key, v))
+        } else {
```
```suggestion
        } else if (value) {
```

`value` is possibly undefined; this probably needs an additional `else` to handle this case.

BernardoSousa03 replied:
I though that giving an undefined value into `validateHeaderValue` would make it raise the error with code `ERR_INVALID_HEADER_VALUE`, since this is the expected error in the Node docs. After some further research i found out that typescript expects the value type to be string, therefore creating an error. 

I will be raising a warning when the value is undefined, but i realised there's also an invalid character error. I think adding it to the `PROXY_ENCOUNTERED_INVALID_HEADER_VALUE` error fits perfectly to the category of the error, since there's the original error attached, letting the user know if its either an undefined value or an invalid character.

The outcome of my changes are the following:
```javascript
// Filter for invalid headers  
  const filteredHeaders = Object.fromEntries(
    Object.entries(headers).filter(([key, value]) => {
      try {
        validateHeaderName(key)
        if (Array.isArray(value)) {
          value.forEach((v) => validateHeaderValue(key, v))
        } else if (value !== undefined) {
          validateHeaderValue(key, value)
        } else {
          error.warning('PROXY_ENCOUNTERED_INVALID_HEADER_VALUE', { [key]: value }, this.req.method, this.req.originalUrl, new TypeError('Header value is undefined while expecting string'));
          return false;
        }
        return true
      } catch (err) {
        if (err.code === 'ERR_INVALID_HTTP_TOKEN') {
          error.warning('PROXY_ENCOUNTERED_INVALID_HEADER_NAME', { [key]: value }, this.req.method, this.req.originalUrl, err)
        } else if (err.code === 'ERR_INVALID_CHAR') {
          error.warning('PROXY_ENCOUNTERED_INVALID_HEADER_VALUE', { [key]: value }, this.req.method, this.req.originalUrl, err)
        } else {
            // rethrow any other errors
            throw err;
        }
        return false
      }
    }),
  )
```
  
 I have also ran the system-tests by using the `yarn test` command in the root of cypress and had no errors. If there's something i'm missing let me know!

cacieprins replied:
That looks like a great solution! Feel free to update this PR with that change, so that CI will pass the `check-ts` step. You'll also need to update the changelog at this point.

Thank you!

cacieprins reviewed (changes requested):
It looks like there's a ts error in `response-middleware.ts` on L#321 - verify by running `yarn check-ts` from the root of the project.  Other than that, looks good. I'm a little concerned this may invalidate some [system test](https://github.com/cypress-io/cypress/tree/develop/system-tests) snapshots - can you run those to verify?

BernardoSousa03 committed:
fix: added undefined verification and catched missing error

Fixed a typescript error where validateHeaderValue was being called
with value possibly being undefined. Fixed catching missing error
where code is 'ERROR_INVALID_CHAR' and rethrows other errors

jennifer-shehane committed:
Merge branch 'develop' into issue-28865

cacieprins committed:
Merge branch 'develop' into issue-28865

cacieprins committed:
Merge branch 'develop' into issue-28865

AtofStryker commented on cli/CHANGELOG.md:
```diff
@@ -1,4 +1,5 @@
 <!-- See the ../guides/writing-the-cypress-changelog.md for details on writing the changelog. -->
+
```
```suggestion
```

AtofStryker replied:
to make the semantic changelog job pass

AtofStryker committed:
Update cli/CHANGELOG.md
